src/irmin-pack/unix/ext.ml

```diff
@@ -218,44 +211,242 @@ module Maker (Config : Conf.S) = struct
+            | pid ->
+                t.during_gc <- Some { next_generation; pid; unlink };
+                Exit.add pid;
+                Ok ()
+
+          let copy_latest_newies ~generation ~last_copied_offset ~root t =
+            let open Result_syntax in
+            let ( - ) = Int63.sub in
+            let old_suffix = File_manager.suffix t.fm in
+            let old_end_offset = File_manager.Suffix.end_offset old_suffix in
+            let remaining = old_end_offset - last_copied_offset in
+            let new_suffix_path =
+              Irmin_pack.Layout.V3.suffix ~root ~generation
+            in
+            let* new_suffix = Io.open_ ~path:new_suffix_path ~readonly:false in
+            Gc.transfer ~src:old_suffix ~dst:new_suffix ~off:last_copied_offset
+              ~len:remaining;
+            Io.close new_suffix
+
+          let swap_and_purge ~unlink ~generation ~last_copied_offset t =
+            let open Result_syntax in
+            let* () =
+              File_manager.swap t.fm ~generation ~unlink ~last_copied_offset
+            in
+            (* No need to purge dict here, as it is global to the store. *)
+            (* No need to purge index here. It is global too, but some hashes may
+               not point to valid offsets anymore. Pack_store will just say that
+               such keys are not member of the store. *)
+            Contents.CA.purge_lru t.contents;
+            Node.CA.purge_lru t.node;
+            Commit.CA.purge_lru t.commit;
+            [%log.info "GC: end"];
+            Ok ()
+
+          let gc_errors (status, gc_output) =
+            match (status, gc_output) with
+            | Lwt_unix.WEXITED n, Error (`Gc_process_error str) ->
+                Error (`Gc_process_error (Fmt.str "Exited %d %s" n str))
+            | Lwt_unix.WEXITED n, Error (`Corrupted_gc_result_file str) ->
+                Error (`Corrupted_gc_result_file (Fmt.str "Exited %d %s" n str))
+            | Lwt_unix.WSIGNALED n, Error (`Corrupted_gc_result_file _) ->
+                Error
+                  (`Gc_process_died_without_result_file
+                    (Fmt.str "Signaled %d" n))
+            | Lwt_unix.WSTOPPED n, Error (`Corrupted_gc_result_file _) ->
+                Error
+                  (`Gc_process_died_without_result_file
+                    (Fmt.str "Stopped %d" n))
+            | Lwt_unix.WSIGNALED n, Error (`Gc_process_error str) ->
+                Error (`Gc_process_error (Fmt.str "Signaled %d %s" n str))
+            | Lwt_unix.WSTOPPED n, Error (`Gc_process_error str) ->
+                Error
+                  (`Gc_process_died_without_result_file
+                    (Fmt.str "Stopped %d %s" n str))
+            | Lwt_unix.WEXITED _, Ok _
+            | Lwt_unix.WSIGNALED _, Ok _
+            | Lwt_unix.WSTOPPED _, Ok _ ->
+                assert false
+
+          let finalise ~wait t =
+            let { next_generation; pid; unlink } = Option.get t.during_gc in
```

Comment:
This will fail with an exception if there is no ongoing GC. What we want is to return `Ok false`.